gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,203 @@
+    // references).
+    // Else if the commit is head commit, add contents to live set.
+    // Else check if the content is expired or not based on cutoff time of this reference.
+    //     Based on that add content to respective sets (live or expired).
+    //
+    // Note: So, If a table is dropped in a reference. Then independent of cutoff time,
+    //     all the snapshots for that table in the reference (that are not live in other reference)
+    // will be
+    //         considered for expiry.
+
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    for (Reference reference : references) {
+      Instant cutOffTimestamp = this.cutOffTimestamp.apply(reference);
+      walkReference(
+          contentValuesCollector,
+          reference,
+          commitMeta ->
+              // If the commit time is older than (think: less than, before) cutoff-time, then
+              // commit is expired.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) < 0);
+    }
+    return new GCResult<>(contentValuesCollector.contentValues);
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void walkReference(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      Predicate<CommitMeta> expiredCommitPredicate) {
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+
+      Set<ContentKey> deletedKeys = new HashSet<>();
+      AtomicBoolean commitHead = new AtomicBoolean(true);
+      commits.forEach(
+          (logEntry) -> {
+            boolean isExpired;
+            if (commitHead.get()) {
+              // If the commit is head commit, then it should be kept live if the table is reachable
+              isExpired = false;
+              commitHead.set(false);
+            } else {
+              // check expiry based on cutoff time
+              isExpired = expiredCommitPredicate.test(logEntry.getCommitMeta());
+            }
+
+            if (logEntry.getOperations() != null) {
+              logEntry
+                  .getOperations()
+                  .forEach(
+                      operation -> {
+                        if (operation instanceof ImmutableDelete) {
+                          deletedKeys.add(operation.getKey());
+                        }
+                      });
+            }
+            handleCommit(
+                contentValuesCollector,
+                reference,
+                logEntry.getCommitMeta(),
+                deletedKeys,
+                isExpired);
+          });
+    } catch (RuntimeException e) {
+      if (e.getCause() instanceof NessieNotFoundException) {
+        LOGGER.info("Reference {} to retrieve commits is no longer exists", reference);
```

Comment:
```suggestion
        LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
```

Reply:
is this the only thing we have to do? A missing reference can be completely collected right?